File: src/p0042/java/solution.java
class Solution {
    public int trap(int[] height) {
        int water = 0, lbound = 0, rbound = 0;
        int left = 0, right = height.length;
        while (left < right) {
            if (lbound < rbound) {
                int curr = height[left++];
                lbound = Math.max(lbound, curr);
                water += lbound - curr;
            } else {
                int curr = height[--right];
                rbound = Math.max(rbound, curr);
                water += rbound - curr;
            }
        }
        return water;
    }
}

File: src/p0042/golang/solution.go
func trap(height []int) int {
	var water, lbound, rbound int
	left, right := 0, len(height)
	for left < right {
		if lbound < rbound {
			curr := height[left]
			left++
			if lbound < curr {
				lbound = curr
			} else {
				water += lbound - curr
			}
		} else {
			right--
			curr := height[right]
			if rbound < curr {
				rbound = curr
			} else {
				water += rbound - curr
			}
		}
	}
	return water
}
